cluster-autoscaler/core/scale_up.go

```diff
@@ -77,8 +78,26 @@ func ScaleUp(context *AutoscalingContext, unschedulablePods []*apiv1.Pod, nodes
 	podsRemainUnschedulable := make(map[*apiv1.Pod]bool)
 	expansionOptions := make([]expander.Option, 0)
 
-	for _, nodeGroup := range context.CloudProvider.NodeGroups() {
+	nodeGroups := context.CloudProvider.NodeGroups()
 
+	if context.AutoscalingOptions.NodeAutoprovisioningEnabled {
+		machines, err := context.CloudProvider.GetAvilableMachineTypes()
+		if err != nil {
+			glog.Warningf("Failed to get machine types: %v", err)
+		} else {
+			bestLabels := labels.BestLabelSet(unschedulablePods)
+			for _, machineType := range machines {
+				nodeGroup, err := context.CloudProvider.NewNodeGroup(machineType, bestLabels, nil)
+				if err != nil {
+					glog.Warningf("Unable to build temporary node group for %s: %v", machineType, err)
+				} else {
+					nodeGroups = append(nodeGroups, nodeGroup)
+				}
+			}
+		}
+	}
```

Comment:
ScaleUp is complex enough as it is. Can you move this whole block to a separate function?

Reply:
Done.